packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -308,12 +326,62 @@ export function dispatchEventForPluginEventSystem(
   );
 }
 
+function getNearestRootOrPortalContainer(instance: Element): Element {
+  let node = getClosestInstanceFromNode(instance);
+  while (node !== null) {
+    const tag = node.tag;
+    // Once we encounter a host container or root container
+    // we can return their DOM instance.
+    if (tag === HostRoot || tag === HostPortal) {
+      return node.stateNode.containerInfo;
+    }
+    node = node.return;
+  }
+  return instance;
+}
+
 export function attachElementListener(listener: ReactDOMListener): void {
-  // TODO
+  const {event, target} = listener;
+  const {passive, type} = event;
+  let containerEventTarget = target;
+  // If we the target is a managed React element, then we need to
+  // find the nearest root/portal contained to attach the event listener
+  // to. If it's not managed, i.e. the window, then we just attach
+  // the listener to the target.
+  const possibleManagedTarget = ((target: any): Element);
+  if (getClosestInstanceFromNode(possibleManagedTarget)) {
+    containerEventTarget = getNearestRootOrPortalContainer(
+      possibleManagedTarget,
+    );
+  }
+  const listenerMap = getListenerMapForElement(containerEventTarget);
+  // Add the event listener to the target container (falling back to
+  // the target if we didn't find one).
+  listenToTopLevelEvent(
+    ((type: any): DOMTopLevelEventType),
+    containerEventTarget,
+    listenerMap,
+    passive,
+  );
+  // Get the internal listeners Set from the target instance.
+  let listeners = getListenersFromTarget(target);
+  // If we don't have any listeners, then we need to init them.
+  if (listeners === null) {
+    listeners = new Set();
+    initListenersSet(target, listeners);
```

Comment:
Curiosity question: Why does `getListenerMapForElement` lazily initialize the Map but `initListenersSet` doesn't?

Reply:
`getListenerMapForElement` is something that has existed for a while. I avoiding lazily initializing the listener Set because we'll be doing a lot of lookups on various instances where they have no `useEvent` listeners, and creating and setting up a Set on those is far too much overhead when they'll likely never have `useEvent` listener Sets.